File: life.go
package main

import (
	"fmt"
	"strings"
	"time"
)

// Rules
const (
	D = iota // Die
	S = iota // Stay the same
	G = iota // Grow a new cell
)

type Life struct {
	generation int
	grid       [][]bool
	height     int
	width      int
	rules      [9]int
}

func makeGrid(width int, height int) [][]bool {
	var grid = make([][]bool, width)
	for i := 0; i < height; i++ {
		grid[i] = make([]bool, height)
	}
	return grid
}

func NewLife(width int, height int, rules [9]int) Life {
	var game = Life{
		generation: 0,
		grid:       makeGrid(width, height),
		height:     height,
		width:      width,
		rules:      rules,
	}
	return game
}

func (l *Life) get(x int, y int) bool {
	if x >= 0 && x < l.width && y >= 0 && y < l.height {
		return l.grid[x][y]
	}
	return false
}

func (l *Life) getNeighbors(x int, y int) int {
	neighbors := 0
	for xOffset := -1; xOffset <= 1; xOffset++ {
		for yOffset := -1; yOffset <= 1; yOffset++ {
			// Don't count the offset 0, 0
			if xOffset != 0 || yOffset != 0 {
				if l.get(x+xOffset, y+yOffset) {
					neighbors++
				}
			}
		}
	}
	return neighbors
}

func (l *Life) printGrid() {
	fmt.Printf("Generation %d\n", l.generation)
	fmt.Printf("+" + strings.Repeat("-", l.width) + "+\n")
	for y := 0; y < l.height; y++ {
		rowText := "|"
		for x := 0; x < l.height; x++ {
			if l.grid[x][y] {
				rowText += "0"
			} else {
				rowText += " "
			}
		}
		rowText += "|\n"
		fmt.Printf(rowText)
	}
	fmt.Printf("+" + strings.Repeat("-", l.width) + "+\n")
}

func (l *Life) nextGeneration() {
	var nextGrid = makeGrid(l.width, l.height)
	for x := 0; x < l.width; x++ {
		for y := 0; y < l.height; y++ {
			neighbors := l.getNeighbors(x, y)
			nextState := l.rules[neighbors]
			if nextState == G {
				nextGrid[x][y] = true
			} else if nextState == S {
				nextGrid[x][y] = l.grid[x][y]
			}
		}
	}
	l.grid = nextGrid
	l.generation++
}

func (l *Life) addFigure(x int, y int, points [][2]int) {
	for _, point := range points {
		l.grid[x+point[0]][y+point[1]] = true
	}
}

var RPENTOMINO = [][2]int{{1, 0}, {2, 0}, {0, 1}, {1, 1}, {1, 2}}
var BLOCK = [][2]int{{0, 0}, {1, 0}, {0, 1}, {1, 1}}
var BLINKER = [][2]int{{1, 0}, {1, 1}, {1, 2}}
var BEACON = [][2]int{{0, 0}, {0, 1}, {1, 0}, {2, 3}, {3, 2}, {3, 3}}

func main() {
	rules := [9]int{D, D, S, G, D, D, D, D, D}
	game := NewLife(10, 10, rules)
	// game.addFigure(0, 0, BEACON)
	game.addFigure(4, 4, RPENTOMINO)

	for {
		game.printGrid()
		game.nextGeneration()
		time.Sleep(time.Second/2)
	}
}

File: life.zig
const std = @import("std");
const warn = std.debug.warn;

// D: Die, S: Stay the same, G: Grow a new cell
const Rule = enum {
    D,  // Die
    S,  // Stay the same
    G,  // Grow a new cell
};

const Life = struct {
    generation: i32,
    rules: []const Rule,
    width: usize,
    height: usize,
    grid: [][]bool,
    next_grid: [][]bool,
};

fn get(self: *Life, x: i32, y: i32) bool {
    if (x >= 0 and y >= 0) {
        const x_usize = @intCast(usize, x);
        const y_usize = @intCast(usize, y);
        if (x_usize < self.width and y_usize < self.height) {
            return self.grid[x_usize][y_usize];
        }
    }
    return false;
}

fn getNeighbors(self: *Life, x: usize, y: usize) usize {
    const x_i32 = @intCast(i32, x);
    const y_i32 = @intCast(i32, y);
    var neighbors: usize = 0;
    for ([_]i32{-1, 0, 1}) |x_offset| {
        for ([_]i32{-1, 0, 1}) |y_offset| {
            // Don't count the offset 0, 0
            if (x_offset != 0 or y_offset != 0) {
                if (get(self, x_i32 + x_offset, y_i32 + y_offset)) {
                    neighbors += 1;
                }
            }
        }
    }
    return neighbors;
}

fn printGrid(self: *Life) void {
    warn("Generation {}\n", self.generation);
    warn("+");
    var i: usize = 0;
    while (i < self.width) {
        warn("-");
        i += 1;
    }
    warn("+\n");
    for (self.grid) |row, y| {
        warn("|");
        for (row) |cell, x| {
            if (self.grid[x][y]) {
                warn("0");
            } else {
                warn(" ");
            }
        }
        warn("|\n");
    }
    warn("+");
    i = 0;
    while (i < self.width) {
        warn("-");
        i += 1;
    }
    warn("+\n");
}

fn nextGeneration(self: *Life) void {
    for (self.grid) |row, y| {
        for (row) |cell, x| {
            const neighbors = getNeighbors(self, x, y);
            const next_state = self.rules[neighbors];
            if (next_state == Rule.G) {
                self.next_grid[x][y] = true;
            } else if (next_state == Rule.D) {
                self.next_grid[x][y] = false;
            }
        }
    }
    self.generation += 1;
    for (self.next_grid) |row, y| {
        for (row) |cell, x1| {
            self.grid[x1][y] = self.next_grid[x1][y];
        }
    }
}

fn addFigure(self: *Life, x: usize, y: usize, points: []const [2]usize) void {
    for (points) |point| {
        self.grid[x+point[0]][y+point[1]] = true;
        self.next_grid[x+point[0]][y+point[1]] = true;
    }
}

const rpentomino = [_][2]usize{
    [2]usize{1, 0},
    [2]usize{2, 0},
    [2]usize{0, 1},
    [2]usize{1, 1},
    [2]usize{1, 2},
};
const block = [_][2]usize{
    [2]usize{0, 0},
    [2]usize{1, 0},
    [2]usize{0, 1},
    [2]usize{1, 1},
};
const blinker = [_][2]usize{
    [2]usize{1, 0},
    [2]usize{1, 1},
    [2]usize{1, 2},
};
const beacon = [_][2]usize{
    [2]usize{0, 0},
    [2]usize{0, 1},
    [2]usize{1, 0},
    [2]usize{2, 3},
    [2]usize{3, 2},
    [2]usize{3, 3},
};

pub fn main() void {
    const rules = [_]Rule{
        Rule.D, Rule.D, Rule.S,
        Rule.G, Rule.D, Rule.D,
        Rule.D, Rule.D, Rule.D
    }; 
    const width = 10; 
    const height = 10;

    // Create grids of slices for current and next generation
    var grid = init: {
        var original_grid: [width*height]bool = [_]bool{false} ** (width*height);
        var sliced_grid: [width][]bool = undefined;
        for (sliced_grid) |*pt, i| {
            pt.* = original_grid[i*height..(i+1)*height];
        }
        break :init sliced_grid[0..sliced_grid.len];
    };
    var next_grid = init: {
        var original_grid: [width*height]bool = [_]bool{false} ** (width*height);
        var sliced_grid: [width][]bool = undefined;
        for (sliced_grid) |*pt, i| {
            pt.* = original_grid[i*height..(i+1)*height];
        }
        break :init sliced_grid[0..sliced_grid.len];
    };

    var game = Life {
        .generation = 0,
        .rules = rules,
        .width = width,
        .height = height,
        .grid = grid,
        .next_grid = next_grid
    };
    
    // addFigure(&game, 0, 0, beacon[0..BEACON.len]);
    addFigure(&game, 4, 4, rpentomino[0..rpentomino.len]);

    while (true) {
        printGrid(&game);
        nextGeneration(&game);
        std.time.sleep(1000000000);
    }
}
